app/src/main/java/com/github/gotify/service/GotifyMessengerService.kt

```diff
@@ -0,0 +1,176 @@
+class GotifyMessengerService : Service() {
+    /** Keeps track of all current registered clients.  */
+    private val db = MessagingDatabase(this)
+    private lateinit var settings: Settings
+
+    /**
+     * Handler of incoming messages from clients.
+     */
+    internal inner class gHandler : Handler() {
+
+        override fun handleMessage(msg: Message) {
+            when (msg.what) {
+                MSG_START -> simpleAnswer(msg, MSG_START)
+                MSG_REGISTER_CLIENT -> {
+                    val uid = msg.sendingUid
+                    val msgData = msg.data
+                    val thread = object : Thread() {
+                        override fun run() {
+                            registerApp(msgData, uid)
+                        }
+                    }
+                    thread.start()
+                    // We need to wait until the registration is done to send url & token :
+                    thread.join()
+                    sendInfo(msg)
+                }
+                MSG_UNREGISTER_CLIENT -> {
+                    val thread = object : Thread() {
+                        val uid = msg.sendingUid
+                        val msgData = msg.data
+                        override fun run() {
+                            unregisterApp(msgData, uid)
+                        }
+                    }
+                    thread.start()
+                    simpleAnswer(msg, MSG_UNREGISTER_CLIENT)
+                }
+                else -> super.handleMessage(msg)
+            }
+        }
+
+        private fun simpleAnswer(msg: Message, what: Int) {
+            try {
+                msg.replyTo?.send(Message.obtain(null, what, 0, 0))
+            } catch (e: RemoteException) {
+            }
+        }
+
+        private fun unregisterApp(msg: Bundle, clientUid: Int) {
+            val clientPackageName = msg.getString("package").toString()
+            // we only trust unregistered demands from the uid who registered the app
+            if (db.strictIsRegistered(clientPackageName, clientUid)) {
+                Log.i("Unregistering $clientPackageName uid: $clientUid")
+                deleteApp(clientPackageName)
+                db.unregisterApp(clientPackageName, clientUid)
+            }
+        }
+
+        private fun registerApp(msg: Bundle,clientUid: Int) {
+            val clientPackageName = msg.getString("package").toString()
+            if (clientPackageName.isBlank()) {
+                Log.w("Trying to register an app without packageName")
+                return
+            }
+            Log.i("registering $clientPackageName uid: $clientUid")
+            // The app is registered with the same uid : we re-register it
+            // the client may need to create a new app in the server
+            if (db.strictIsRegistered(clientPackageName, clientUid)) {
+                Log.i("$clientPackageName already registered : unregistering to register again")
```

Comment:
The user of the gotify-android app should manually grant permissions to other applications to receive these messages. Otherwise, this looks like any app can just intercept gotify messages, without the user knowing it.

Reply:
Any app can register but not intercept messages. But it is a really good idea to grant permissions manually. We can do it with a notification maybe

Reply:
Ahh yeah, thought you could subscribe to existing apps. A notification like. XX likes to receive applications from gotify. Allow, Deny sounds great.